File: src/net/reichholf/dreamdroid/helpers/enigma2/DeviceInfo.java
/* © 2010 Stephan Reichholf <stephan at reichholf dot net>
 * 
 * Licensed under the Create-Commons Attribution-Noncommercial-Share Alike 3.0 Unported
 * http://creativecommons.org/licenses/by-nc-sa/3.0/
 */

package net.reichholf.dreamdroid.helpers.enigma2;

import java.util.ArrayList;

import net.reichholf.dreamdroid.dataProviders.SaxDataProvider;
import net.reichholf.dreamdroid.helpers.ExtendedHashMap;
import net.reichholf.dreamdroid.helpers.SimpleHttpClient;
import net.reichholf.dreamdroid.parsers.GenericSaxParser;
import net.reichholf.dreamdroid.parsers.enigma2.saxhandler.E2DeviceInfoHandler;

import org.apache.http.NameValuePair;

public class DeviceInfo {
	public static final String KEY_GUI_VERSION = "guiversion";
	public static final String KEY_IMAGE_VERSION = "imageversion";
	public static final String KEY_INTERFACE_VERSION = "interfaceversion";
	public static final String KEY_FRONT_PROCESSOR_VERSION = "fpversion";
	public static final String KEY_DEVICE_NAME = "devicename";
	public static final String KEY_FRONTENDS = "frontends";
	public static final String KEY_FRONTEND_NAME = "name";
	public static final String KEY_FRONTEND_MODEL = "model";
	public static final String KEY_NICS = "nics";
	public static final String KEY_NIC_NAME = "name";
	public static final String KEY_NIC_MAC = "mac";
	public static final String KEY_NIC_DHCP = "dhcp";
	public static final String KEY_NIC_IP = "ip";
	public static final String KEY_NIC_GATEWAY = "gateway";
	public static final String KEY_NIC_NETMASK = "netmask";
	public static final String KEY_HDDS = "hdds";
	public static final String KEY_HDD_MODEL = "model";
	public static final String KEY_HDD_CAPACITY = "capacity";
	public static final String KEY_HDD_FREE_SPACE = "free";

	/**
	 * @param shc
	 * @return
	 */
	public static String get(SimpleHttpClient shc) {
		if (shc.fetchPageContent(URIStore.DEVICE_INFO, new ArrayList<NameValuePair>())) {
			return shc.getPageContentString();
		}

		return null;
	}

	/**
	 * @param xml
	 * @param map
	 * @return
	 */
	public static boolean parse(String xml, ExtendedHashMap map) {
		SaxDataProvider sdp = new SaxDataProvider(new GenericSaxParser());

		E2DeviceInfoHandler handler = new E2DeviceInfoHandler(map);
		sdp.getParser().setHandler(handler);

		if (sdp.parse(xml)) {
			return true;
		}

		return false;
	}

}

File: app/src/net/reichholf/dreamdroid/enigma/DeviceInfo.kt
package net.reichholf.dreamdroid.enigma

import java.io.Serializable

data class DeviceFrontend(
    val name: String = "",
    val model: String = "",
) : Serializable

data class DeviceNic(
    val name: String = "",
    val mac: String = "",
    val dhcp: String = "",
    val ip: String = "",
    val gateway: String = "",
    val netmask: String = "",
) : Serializable

data class DeviceHdd(
    val model: String = "",
    val capacity: String = "",
    val free: String = "",
) : Serializable

/**
 * Typed `/web/deviceinfo` payload.
 */
data class DeviceInfo(
    val guiVersion: String = "",
    val imageVersion: String = "",
    val interfaceVersion: String = "",
    val frontProcessorVersion: String = "",
    val deviceName: String = "",
    val frontends: List<DeviceFrontend> = emptyList(),
    val nics: List<DeviceNic> = emptyList(),
    val hdds: List<DeviceHdd> = emptyList(),
) : Serializable {
    fun isEmpty(): Boolean {
        return guiVersion.isEmpty()
            && imageVersion.isEmpty()
            && interfaceVersion.isEmpty()
            && frontProcessorVersion.isEmpty()
            && deviceName.isEmpty()
            && frontends.isEmpty()
            && nics.isEmpty()
            && hdds.isEmpty()
    }
}
